pkg/controllers/disruption/controller.go

```diff
@@ -179,194 +163,60 @@ func (c *Controller) disrupt(ctx context.Context, disruption Method) (bool, erro
-	deprovisioningActionsPerformedCounter.With(map[string]string{
-		actionLabel:        fmt.Sprintf("%s/%s", m.Type(), cmd.Action()),
-		deprovisionerLabel: m.Type(),
-	}).Inc()
 	disruptionActionsPerformedCounter.With(map[string]string{
 		actionLabel:            string(cmd.Action()),
 		methodLabel:            m.Type(),
 		consolidationTypeLabel: m.ConsolidationType(),
 	}).Inc()
 	logging.FromContext(ctx).Infof("disrupting via %s %s", m.Type(), cmd)
 
-	reason := fmt.Sprintf("%s/%s", m.Type(), cmd.Action())
-	if cmd.Action() == ReplaceAction {
-		if err := c.launchReplacementNodeClaims(ctx, m, cmd); err != nil {
-			// If we failed to launch the replacement, don't disrupt.  If this is some permanent failure,
-			// we don't want to disrupt workloads with no way to provision new NodeClaims for them.
-			return fmt.Errorf("launching replacement, %w", err)
-		}
+	stateNodes := lo.Map(cmd.candidates, func(c *Candidate, _ int) *state.StateNode {
+		return c.StateNode
+	})
+	// Cordon the old nodes before we launch the replacements to prevent new pods from scheduling to the old nodes
+	if err := state.RequireNoScheduleTaint(ctx, c.kubeClient, true, stateNodes...); err != nil {
+		return multierr.Append(fmt.Errorf("tainting nodes, %w", err), state.RequireNoScheduleTaint(ctx, c.kubeClient, false, stateNodes...))
 	}
 
-	for _, candidate := range cmd.candidates {
-		c.recorder.Publish(disruptionevents.Terminating(candidate.Node, candidate.NodeClaim, reason)...)
-
-		if err := nodeclaimutil.Delete(ctx, c.kubeClient, candidate.NodeClaim); err != nil {
-			if !errors.IsNotFound(err) {
-				logging.FromContext(ctx).Errorf("terminating, %s", err)
-			}
-			continue
+	var nodeClaimKeys []nodeclaim.Key
+	var err error
+	if len(cmd.replacements) > 0 {
+		if nodeClaimKeys, err = c.createReplacementNodeClaims(ctx, m, cmd); err != nil {
+			// If we failed to launch the replacement, don't disrupt.  If this is some permanent failure,
+			// we don't want to disrupt workloads with no way to provision new nodes for them.
+			return multierr.Append(fmt.Errorf("launching replacement nodeclaim, %w", err), state.RequireNoScheduleTaint(ctx, c.kubeClient, false, stateNodes...))
 		}
-		nodeclaimutil.TerminatedCounter(candidate.NodeClaim, reason).Inc()
 	}
 
-	// We wait for NodeClaims to delete to ensure we don't start another round of disruption
-	// until this node is fully deleted.
-	for _, oldCandidate := range cmd.candidates {
-		c.waitForDeletion(ctx, oldCandidate.NodeClaim)
+	providerIDs := lo.Map(cmd.candidates, func(c *Candidate, _ int) string { return c.ProviderID() })
+	// We have the new NodeClaims created at the API server so mark the old NodeClaims for deletion
+	c.cluster.MarkForDeletion(providerIDs...)
+
+	if err := c.queue.Add(orchestration.NewCommand(nodeClaimKeys,
+		lo.Map(cmd.candidates, func(c *Candidate, _ int) *state.StateNode { return c.StateNode }), c.clock.Now(), m.Type(), m.ConsolidationType())); err != nil {
```

Comment:
Shouldn't the queue own the clock?
Why not pass in the whole method rather than deconstructing the types?

Reply:
Good catch, I had left this in as part of a past revision as well.

Reply:
I wanted to keep the Add() parameter as a function so that we could more easily create a command in testing and inject it into the queue, rather than inputting all parameters then having to grab the created command from the testing datastructures.